LPC176x Framework changes

p3p opened the pull request:
### Description
Changes needed to Marlin for compatibility with changes to the LPC176x framework.

### Benefits
Marlin continues to build, also better ADC handing, consistently fast gpio (const and variable) and pins debug shouldn't crash.

Not finalised and merge needs coordinated with framework release, will need to add framework version check, finish pins file changes.

p3p committed:
update

p3p committed:
version checks and misc

p3p committed:
Update LPC176x pins

LinFor commented:
Can't compile Marlin with this. When I use pio-framework-arduino-lpc176x:master branch, I see many errors about undeclared and redeclared symbols:
```
In file included from c:\users\shashlov\documents\arduino\marlin\marlin\src\hal\shared\marduino.h:36,
                 from Marlin\src\HAL\HAL_LPC1768\../../core/../inc/../HAL/./HAL_LPC1768/HAL.h:47,
                 from Marlin\src\HAL\HAL_LPC1768\../../core/../inc/../HAL/HAL.h:26,
                 from Marlin\src\HAL\HAL_LPC1768\../../core/../inc/MarlinConfig.h:30,
                 from Marlin\src\HAL\HAL_LPC1768\../../core/serial.h:24,
                 from Marlin\src\HAL\HAL_LPC1768\DebugMonitor.cpp:26:
C:\Users\Shashlov\.platformio\packages\framework-arduino-lpc176x\cores\arduino/Arduino.h: In function 'void digitalWrite(pin_t, uint8_t)':
C:\Users\Shashlov\.platformio\packages\framework-arduino-lpc176x\cores\arduino/Arduino.h:119:8: error: 'pin_is_valid' was not declared in this scope
   if (!pin_is_valid(pin)) return;
        ^~~~~~~~~~~~
C:\Users\Shashlov\.platformio\packages\framework-arduino-lpc176x\cores\arduino/Arduino.h:119:8: note: suggested alternative: 'pin_status'
   if (!pin_is_valid(pin)) return;
        ^~~~~~~~~~~~
        pin_status
C:\Users\Shashlov\.platformio\packages\framework-arduino-lpc176x\cores\arduino/Arduino.h:121:3: error: 'gpio_set' was not declared in this scope
   gpio_set(pin, pin_status);
   ^~~~~~~~
......
```

When I use pio-framework-arduino-lpc176x:pr_gpio_new branch - It fails on package version, but if I omit this checks - it compiles successfully, but I not sure then it would work as expected.

For which commit in pio-framework-arduino-lpc176x this PR was designed?

p3p commented:
As long as the version check passes for `nxplpc-arduino-lpc176x` and `framework-arduino-lpc176x` they should be the correct versions to build, You should use ` pio-framework-arduino-lpc176x:master`.

p3p committed:
make sure HAL_TIMER_RATE is constexpr

p3p committed:
cleanup

thinkyhead committed:
Move tests to SanityCheck.h

thinkyhead commented:
Whenever the framework is released, this seems good to go.

LinFor commented:
> As long as the version check passes for `nxplpc-arduino-lpc176x` and `framework-arduino-lpc176x` they should be the correct versions to build, You should use ` pio-framework-arduino-lpc176x:master`.

Now I tested it.
Adc works very well, real 12-bit hardware resoluton. Increasing to 14-bit (oversampled of course) exposes some noise, but 12-bit works very stable. I run very quickly between printing, but I has about 0.3 degree resolution with pt100 through amp on 100 degrees, and 0.06 degree resolution on my bed thermistor on 60 degrees. PID regulation is very stable.

But.... Seems software spi is broken with this new framework. At least the TMC2130 drivers failed to work. Current framework on the same configuration (SKR 1.3) works normally.

p3p commented:
You would probably need to increase the low pass filter another bit or 2 to use 14 bit adc but it more than likely isn't worth the latency increase, the 23 slot median filter is unnecessary for most people as well.

I'l look into the tmc software spi issue, the library uses its own software spi implementation.

thinkyhead commented on Marlin/src/HAL/HAL_LPC1768/HAL.h:
```diff
@@ -132,11 +132,40 @@ int freeMemory();
+[[gnu::always_inline]] inline void HAL_start_adc(const pin_t pin) {
+  HAL_adc_reading = FilteredADC::read(pin) >> 6; // returns 16bit value, reduce to 10bit
+}
+[[gnu::always_inline]] inline uint16_t HAL_read_adc() {
```
As far as I know `FORCE_INLINE` is available here.

p3p replied:
Indeed it is, I sometimes forget to switch back to Marlin convention, I tend to avoid the preprocessor by any means necessary.

p3p commented:
@LinFor I've patched an issue with `pinMode` which may fix software spi but it could be running too fast for the drivers / to be stable. Will need to update TMCStepper lib to rate limit it.

p3p commented:
@LinFor have you been able to check if this was all that was required, or if I need to limit the speed?

LinFor commented:
> @LinFor have you been able to check if this was all that was required, or if I need to limit the speed?

I've checked and SCK speed too fast... WIth current released lpc-framework it was arount 1.5MHz:
![image](https://user-images.githubusercontent.com/24222858/68047248-f1c2c400-fcee-11e9-83b6-e786069374b7.png)
WIth this new framework we had around 10MHz:
![image](https://user-images.githubusercontent.com/24222858/68047295-174fcd80-fcef-11e9-9f26-1200a935a4aa.png)

I'd added bogus arduino delays to writeSCK_L and writeSCK_H and it works normally, printing ok on my hardware setup, so seems tmc sw-spi is the only problem blocks this PR.

I think this must be patched into the TMCStepper sw-spi implementation, not general gpio slowdown... @teemuatlut can you help us?

p3p commented:
already discussed it with teemuatlut was just waiting for confirmation it needed slowed down, shouldn't be an issue, I should have predicted there may be some weird consequences by improving the gpio performance.

p3p committed:
namespace update

p3p commented:
@LinFor I added some delay to this branch (https://github.com/p3p/TMCStepper/tree/pr_lpctimingfix) of the TMCStepper lib, can you test if its enough and possibly what the actual frequency is, my cheap logic analyser is struggling.

LinFor commented:
> @LinFor I added some delay to this branch (https://github.com/p3p/TMCStepper/tree/pr_lpctimingfix) of the TMCStepper lib, can you test if its enough and possibly what the actual frequency is, my cheap logic analyser is struggling.

Now all works fine!
Actual SCK freq is around 3.3Mhz. The edges are a bit overwhelmed, but TMC responds stably on my setup.

![image](https://user-images.githubusercontent.com/24222858/68068143-1ce6fb00-fd63-11e9-825f-424644ddbc48.png)

But to successfully build I had to add LPC176x namespace to gpio_set()/delay_ns() in Adafruit_NeoPixel.cpp and pwm_set_frequency()/pwm_write_ratio() in fast_pwm.cpp at HAL

PS: Also I trying change 40ns delay to 30ns, it's works fine on 3.7Mhz:

![image](https://user-images.githubusercontent.com/24222858/68068394-69333a80-fd65-11e9-8335-25eb0c7db376.png)

And to 25ns, 4.2Mhz TMC starts to currupt answers:

![image](https://user-images.githubusercontent.com/24222858/68068474-3e95b180-fd66-11e9-9f3e-68ace4d0eac0.png)

p3p commented:
Thanks for testing, I don't want to push closer to 4 Mhz as `delay_ns(40)` is only 4 nops, at 100Mhz, and on the 120Mhz LPC1769 .. it's 4.8 nops rounded to .. 4 nops because integer division.

Sorry about the compile issue, knew I'd miss some places Marlin used internal framework functions with that namespace update, I've already updated my fork of Adafruit_NeoPixel but not pushed it yet.

p3p committed:
update platformio.ini

Make sure previous versions of Marlin can still compile and use links to temporary hotfix versions of libraries.

thinkyhead commented on Marlin/src/HAL/HAL_LPC1768/pinsDebug.h:
```diff
@@ -33,52 +33,21 @@
-  int8_t pin_mode = get_pin_mode(pin);
-  if (pin_mode == -1 || pin_mode == ADC_pin_mode(pin)) // found an invalid pin or active analog pin
+bool GET_PINMODE(const pin_t pin) {
+  if (!LPC176x::pin_is_valid(pin) || LPC176x::pin_adc_enabled(pin)) // found an invalid pin or active analog pin
```
@p3p — According to #22568 this might have broken `M43` reporting of analog pins, and maybe this has other side-effects. Can you re-validate these changes and see if there is something amiss?
